src/protocols/UBX.cpp

```diff
@@ -0,0 +1,131 @@
+#include <stdint.h>
+#include <stddef.h>
+
+#include <array>
+#include <vector>
+
+#include <gps_ublox/protocols/UBX.hpp>
+
+using namespace std;
+using namespace gps_ublox;
+using namespace protocols;
+
+const uint8_t UBX::SYNC_1 = 0xB5;
+const uint8_t UBX::SYNC_2 = 0x62;
+const uint8_t UBX::SYNC_1_IDX = 0;
+const uint8_t UBX::SYNC_2_IDX = 1;
+const uint8_t UBX::MSG_CLASS_IDX = 2;
+const uint8_t UBX::MSG_ID_IDX = 3;
+const uint8_t UBX::LENGTH_LSB_IDX = 4;
+const uint8_t UBX::LENGTH_MSB_IDX = 5;
+const uint8_t UBX::PAYLOAD_IDX = 6;
+const size_t UBX::FRAMING_SIZE_OVERHEAD = 8;
+
+UBX::Frame UBX::Frame::fromPacket(const uint8_t *buffer, size_t size)
+{
+    Frame frame;
+    frame.msg_class = buffer[MSG_CLASS_IDX];
+    frame.msg_id = buffer[MSG_ID_IDX];
+
+    uint16_t payload_size = buffer[LENGTH_LSB_IDX] | (buffer[LENGTH_MSB_IDX] << 8);
+    frame.payload = vector<uint8_t>(&buffer[PAYLOAD_IDX], &buffer[PAYLOAD_IDX] + payload_size);
+
+    return frame;
+}
+
+vector<uint8_t> UBX::Frame::toPacket() const
+{
+    vector<uint8_t> packet;
```

Comment:
Since the packet size is known in advance, I'd reserve the vector size

It's a different style, but you could also initialize packet with list-initializers

~~~
vector<uint8_t> packet = { SYNC_1, SYNC_2, this->msg_class, ... };
~~~

Reply:
"Could" as in "I would prefer if you did so" ?

Reply:
No. Could as in "do as you please, this is only a suggestion"